11_FFTBloom/main.cpp

```diff
@@ -0,0 +1,1192 @@
+		{
+			IGPUSampler::SParams params =
+			{
+				textureWrap,
+				textureWrap,
+				textureWrap,
+				ISampler::ETBC_FLOAT_OPAQUE_BLACK,
+				ISampler::ETF_LINEAR,
+				ISampler::ETF_LINEAR,
+				ISampler::ESMM_LINEAR,
+				3u,
+				0u,
+				ISampler::ECO_ALWAYS
+			};
+			return m_device->createSampler(std::move(params));
+		};
+
+		smart_refctd_ptr<IGPUPipelineLayout> imageFirstAxisFFTPipelineLayout;
+		{
+			auto sampler = createSampler(ISampler::E_TEXTURE_CLAMP::ETC_MIRROR);
+			IGPUDescriptorSetLayout::SBinding bnd =
+			{
+				IDescriptorSetLayoutBase::SBindingBase(),
+				0u,
+				IDescriptor::E_TYPE::ET_COMBINED_IMAGE_SAMPLER,
+				IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+				IShader::E_SHADER_STAGE::ESS_COMPUTE,
+				1u,
+				&sampler
+			};
+
+			imageFirstAxisFFTPipelineLayout = createPipelineLayout({ &bnd, 1 });
+		}
+
+		smart_refctd_ptr<IGPUPipelineLayout> lastAxisFFT_convolution_lastAxisIFFTPipelineLayout;
+		{
+			auto sampler = createSampler(ISampler::E_TEXTURE_CLAMP::ETC_MIRROR);
+			smart_refctd_ptr<IGPUSampler> samplers[CHANNELS];
+			std::fill_n(samplers, CHANNELS, sampler);
+			IGPUDescriptorSetLayout::SBinding bnd =
+			{
+				IDescriptorSetLayoutBase::SBindingBase(),
+				0u,
+				IDescriptor::E_TYPE::ET_COMBINED_IMAGE_SAMPLER,
+				IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+				IShader::E_SHADER_STAGE::ESS_COMPUTE,
+				CHANNELS,
+				samplers
+			};
+
+			lastAxisFFT_convolution_lastAxisIFFTPipelineLayout = createPipelineLayout({ &bnd, 1 });
+		}
+
+		smart_refctd_ptr<IGPUPipelineLayout> imageFirstAxisIFFTPipelineLayout;
+		{
+			IGPUDescriptorSetLayout::SBinding bnd =
+			{
+				IDescriptorSetLayoutBase::SBindingBase(),
+				0u,
+				IDescriptor::E_TYPE::ET_STORAGE_IMAGE,
+				IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+				IShader::E_SHADER_STAGE::ESS_COMPUTE,
+				1,
+				nullptr
+			};
+
+			imageFirstAxisIFFTPipelineLayout = createPipelineLayout({ &bnd, 1 });
+		}
+
+		m_readCache = nullptr;
+		m_writeCache = core::make_smart_refctd_ptr<IShaderCompiler::CCache>();
+		auto shaderCachePath = localOutputCWD / "cache.bin";
+
+		{
+			core::smart_refctd_ptr<system::IFile> shaderReadCacheFile;
+			{
+				system::ISystem::future_t<core::smart_refctd_ptr<system::IFile>> future;
+				m_system->createFile(future, shaderCachePath.c_str(), system::IFile::ECF_READ);
+				if (future.wait())
+				{
+					future.acquire().move_into(shaderReadCacheFile);
+					if (shaderReadCacheFile)
+					{
+						const size_t size = shaderReadCacheFile->getSize();
+						if (size > 0ull)
+						{
+							std::vector<uint8_t> contents(size);
+							system::IFile::success_t succ;
+							shaderReadCacheFile->read(succ, contents.data(), 0, size);
+							if (succ)
+								m_readCache = IShaderCompiler::CCache::deserialize(contents);
+						}
+					}
+				}
+				else
+					m_logger->log("Failed Opening Shader Cache File.", ILogger::ELL_ERROR);
+			}
+
+		}
```

Comment:
because shader compilation is long and annoying, I often compile shaders first, before loading any assets